### PR TITLE
[WebUI] Add QuickPhrasesComponent, MiniBarComponent, and related app states

### LINKED ISSUE #165: [WebUI] Allow quick navigation between the quick-phrase mode and the abbreviation expansion mode
AE can be represented as a left-nav icon just like the icons for the quick-phrases.

### PR DESCRIPTION
**Pull request by caisq**
Fixes #165.

The `QuickPhrasesComponent` is composed of `PhraseComponent`s that are controlled by gaze-clickable scroll-up and scroll-down buttons. See screenshot: 
image

Currently, the phrases input to `QuickPhrasesComponent` instances are hardcoded in the `AppComponent`. Follow-up PRs will retrieve them from a remote API endpoint. 

The `AppComponent` now has a number of AppState enum values controlling its overall state. See the left side of the screenshot above for the left nav buttons that allows the user to switch between the different quick phrase sets and the abbreviation expansion mode. The button at the bottom-left corner minimizes the app to a gaze-clickable button that can expand the app back to the previous non-minimized state. See screenshot:
image

These buttons are all gaze-clickable.

Related changes in this PR:
- Unit tests are added for the new components: `QuickPhrasesComponent` and `MiniBarCompoent`
- Unit tests are updated for the existing component: `AppComponent`
- In `ExternalEventsComponent`, allow unregistering of previously registered keypress callbacks. This is now necessary because `AbbreviationComponent` can now get destroyed and re-created in the `AppComponent` depending on its state. So we want to avoid leaking previous callback registrations.
- In cefsharp.ts API, allow specification of a container (parent) view port (DOMRect), so that elements that are invisible due to scroll state can be skipped when registering gaze-clickable button boxes. This is used in `QuickPhrasesComponent` when there are many elements causing overflow and scrolling. See screenshot above.

### CODE REVIEW

File: webui/src/app/app.component.ts [hunk 1]
```diff
@@ -134,4 +197,81 @@ export class AppComponent implements OnInit, AfterViewInit {
...
+      case AppState.QUICK_PHRASES_CARE:
+        return `/assets/images/quick-phrases-care-${activeStateString}.png`;
+      case AppState.ABBREVIATION_EXPANSION:
+        return `/assets/images/abbreviation-expansion-${activeStateString}.png`;
+      default:
+        throw new Error(`Invalid app state: ${this.appState}`);
+    }
+  }
+
+  // TODO(cais): Do not hardcode. Query these from the SpeakFasterService
```
Comment (antonkast-google) [inline webui/src/app/app.component.ts:234]: Am I supposed to know how the phrases got into the SpeakFasterService?

File: webui/src/utils/cefsharp.ts [hunk 2]
```diff
@@ -29,22 +29,36 @@ export async function bindCefSharpListener() {
  * @param instanceId Unique identifier for the component instance. Different
  *   instances of the same component type must have different `instanceId`s.
  * @param elements The list of clickable buttons to register.
+ * @param containerRect Optional DOMRect object that specifies the container
+ *   rectangle. If provided, it will cause only the elements that are
+ *   partially or entirely visible in the containerRect to be registered.
+ *   If not provided (default), all elements will be reigstered regardless
+ *   of location.
```
Comment (antonkast-google) [inline webui/src/utils/cefsharp.ts:36]: Is this supposed to account for the problem of windows obscuring the native buttons? It seems to me the visible area is not necessarily rectangular.

**Review by antonkast-google — APPROVED** [review]

File: webui/src/app/app.component.ts (hunk 1 above)
Comment (caisq) [reply webui/src/app/app.component.ts:234]: Let's follow up on that discussion internally.

File: webui/src/utils/cefsharp.ts (hunk 2 above)
Comment (caisq) [reply webui/src/utils/cefsharp.ts:36]: This mechanism is not designed for that purpose. It is design for the case where you have a `div` element containing children elements inside and the outer `div` element has `overflow: scroll`. The interaction with other applications, which is what you are asking about IIUC, will be taken care of at level of the C# host app.

**Comment by caisq** [comment]
Thanks for the review, @antonkastgoogle.

**Pull request merged**